java/code/src/com/suse/manager/webui/controllers/clusters/ClustersController.java

```diff
@@ -274,24 +284,109 @@ private static ModelAndView showClusterUpgrade(Request request, Response respons
+         */
+        public String getDescription() {
+            return description;
+        }
+
+        /**
+         * @return provider to get
+         */
+        public String getProvider() {
+            return provider;
+        }
+
+        /**
+         * @return managementNodeId to get
+         */
+        public Long getManagementNodeId() {
+            return managementNodeId;
+        }
+
+        /**
+         * @return managementSettings to get
+         */
+        public Map<String, Object> getManagementSettings() {
+            return managementSettings;
+        }
+    }
+
     private static String addCluster(Request request, Response response, User user) {
-        Optional<Map<String, Object>> json = parseJson(request, response);
-        if (json.isEmpty()) {
+        AddClusterRequest clusterRequest;
+        try {
+            clusterRequest = GSON.fromJson(request.body(), AddClusterRequest.class);
+        }
+        catch (JsonParseException e) {
+            LOG.error("Error parsing request body", e);
             return json(response, HttpStatus.SC_BAD_REQUEST,
-                    ResultJson.error("request_error"));
+                    ResultJson.error("invalid_request_body"));
+        }
+
+        if (StringUtils.isBlank(clusterRequest.getName())) {
+            return json(response, HttpStatus.SC_BAD_REQUEST,
+                    ResultJson.error("Name is mandatory"));
+        }
+        if (StringUtils.isBlank(clusterRequest.getLabel())) {
+            return json(response, HttpStatus.SC_BAD_REQUEST,
+                    ResultJson.error("Label is mandatory"));
+        }
+
+        if (!LABEL_REGEX.matcher(clusterRequest.getLabel()).matches()) {
+            return json(response, HttpStatus.SC_BAD_REQUEST,
+                    ResultJson.error("Label format is wrong"));
+        }
+        if (StringUtils.isBlank(clusterRequest.getDescription())) {
+            return json(response, HttpStatus.SC_BAD_REQUEST,
+                    ResultJson.error("Description is mandatory"));
+        }
+        if (StringUtils.isBlank(clusterRequest.getProvider())) {
+            return json(response, HttpStatus.SC_BAD_REQUEST,
+                    ResultJson.error("Provider is mandatory"));
+        }
+        if (!validProvider(clusterRequest.getProvider())) {
+            return json(response, HttpStatus.SC_BAD_REQUEST,
+                    ResultJson.error("Invalid provider"));
+        }
+        if (clusterRequest.getManagementNodeId() == null) {
+            return json(response, HttpStatus.SC_BAD_REQUEST,
+                    ResultJson.error("Management node id is mandatory"));
+        }
+        if (clusterRequest.getManagementSettings() == null ||
+                clusterRequest.getManagementSettings().isEmpty()) {
+            return json(response, HttpStatus.SC_BAD_REQUEST,
+                    ResultJson.error("Management settings are mandatory"));
         }
 
-        // TODO validate input
```

Comment:
We should also make sure that the name/label does not exist (both here and in React).
Otherwise, the process reports a confusing error about a duplicate system group name. We should explicitly tell the user that a cluster with this name already exists.

Reply:
done